src/hotspot/share/runtime/handshake.cpp

```diff
@@ -276,6 +280,22 @@ class VM_HandshakeAllThreads: public VM_Handshake {
   VMOp_Type type() const { return VMOp_HandshakeAllThreads; }
 };
 
+void HandshakeOperation::prepare(JavaThread* current_target, Thread* executing_thread) {
+  if (current_target->is_terminated()) {
+    // Will never execute any handshakes on this thread.
+    return;
+  }
+  if (current_target != executing_thread) {
+    // Only when the target is not executing the handshake itself.
+    StackWatermarkSet::start_processing(current_target, StackWatermarkKind::gc);
+  }
```

Comment:
As per my comment to Erik, this situation seems like a programming error to me as there you should not be a way for the target to create a Handle that the executing thread then gets a hold off - data should only flow one way from the requesting thread to the executing thread.

Reply:
StackWatermarkSet::start_processing(current_target, StackWatermarkKind::gc); is for thread local resources which may contain oop. A handshake may read from e.g. the target threads stack, so before the executing thread reads from the stack, we make sure it is safe.